python_modules/dagster/dagster/_core/executor/step_delegating/step_delegating_executor.py

```diff
@@ -68,24 +69,44 @@ def __init__(
             ),
         )
         self._should_verify_step = should_verify_step
+
         self._event_cursor: Optional[str] = None
+        self._pop_events_offset = int(os.getenv("EXECUTOR_POP_EVENTS_OFFSET", "0"))
 
     @property
     def retries(self):
         return self._retries
 
-    def _pop_events(self, instance: DagsterInstance, run_id: str) -> Sequence[DagsterEvent]:
+    def _pop_events(
+        self, instance: DagsterInstance, run_id: str, seen_storage_ids: Set[int]
+    ) -> Sequence[DagsterEvent]:
+        adjusted_cursor = self._event_cursor
+
+        if self._pop_events_offset > 0 and self._event_cursor:
+            cursor_obj = EventLogCursor.parse(self._event_cursor)
+            check.invariant(
+                cursor_obj.is_id_cursor(),
+                "Applying a tailer offset only works with an id-based cursor",
+            )
+            adjusted_cursor = EventLogCursor.from_storage_id(
+                cursor_obj.storage_id() - self._pop_events_offset
+            ).to_string()
+
         conn = instance.get_records_for_run(
             run_id,
-            self._event_cursor,
+            adjusted_cursor,
             of_type=set(DagsterEventType),
         )
         self._event_cursor = conn.cursor
+
         dagster_events = [
             record.event_log_entry.dagster_event
             for record in conn.records
-            if record.event_log_entry.dagster_event
+            if record.event_log_entry.dagster_event and record.storage_id not in seen_storage_ids
         ]
+
+        seen_storage_ids.update([record.storage_id for record in conn.records])
```

Comment:
nit: drop the list instantiation `[` and just feed the comprehension iterator in directly